File: app/load/rearrangement_load.py
#
# Import data into AIRR mongo repository for a public project. This is for
# the repertoire metadata. This assumes you are running in the docker container.
#
# This is a partial hack because currently VDJServer does not produce AIRR TSV
# files with repertoire_id's and data_processing_id's assigned to the rearrangements.
#
# This script assumes that all the rearrangements for a single repertoire are
# in a single file. It also assumes that data_processing_files is
# set in the primary data_processing.
#

import json
from dotenv import load_dotenv
import os
import sys
import airr
import yaml
import requests
import argparse

# Setup
def getConfig():
    if load_dotenv(dotenv_path='/api-js-tapis/.env'):
        cfg = {}
        cfg['api_server'] = os.getenv('WSO2_HOST')
        cfg['api_key'] = os.getenv('WSO2_CLIENT_KEY')
        cfg['api_secret'] = os.getenv('WSO2_CLIENT_SECRET')
        cfg['username'] = os.getenv('VDJ_SERVICE_ACCOUNT')
        cfg['password'] = os.getenv('VDJ_SERVICE_ACCOUNT_SECRET')
        cfg['dbname'] = os.getenv('MONGODB_DB')
        return cfg
    else:
        print('ERROR: loading config')
        return None

# Fetches a user token based on the supplied auth object
# and returns the auth object with token data on success
def getToken(config):
    data = {
        "grant_type":"password",
        "scope":"PRODUCTION",
        "username":config['username'],
        "password":config['password']
    }
    headers = {
        "Content-Type":"application/x-www-form-urlencoded"
    }

    url = 'https://' + config['api_server'] + '/token'

    resp = requests.post(url, data=data, headers=headers, auth=(config['api_key'], config['api_secret']))
    token = resp.json()
    return token

# Delete all rearrangements from a load set for the repertoire_id
def deleteLoadSet(token, config, repertoire_id, load_set):
    headers = {
        "Content-Type":"application/json",
        "Accept": "application/json",
        "Authorization": "Bearer " + token['access_token']
    }

    # delete rearrangements for given repertoire_id
    if load_set == 0:
        url = 'https://' + config['api_server'] + '/meta/v3/' + config['dbname'] + '/rearrangement/*?filter=' + requests.utils.quote('{"repertoire_id":"' + repertoire_id + '"}')
    else:
        url = 'https://' + config['api_server'] + '/meta/v3/' + config['dbname'] + '/rearrangement/*?filter=' + requests.utils.quote('{"repertoire_id":"' + repertoire_id + '","vdjserver_load_set":' + str(load_set) + '}')
    print(url)
    resp = requests.delete(url, headers=headers)
    print(resp.json())

# Insert the rearrangements for a repertoire
def insertRearrangement(token, config, records):
    # get a token each time, in case it expired
    token = getToken(config)

    headers = {
        "Content-Type":"application/json",
        "Accept": "application/json",
        "Authorization": "Bearer " + token['access_token']
    }

    # insert the rearrangement
    url = 'https://' + config['api_server'] + '/meta/v3/' + config['dbname'] + '/rearrangement/'
    #data = [ record ]
    resp = requests.post(url, json=records, headers=headers)
    data = resp.json()
    if data.get('inserted'):
        print("Inserted records: " + str(data['inserted']))
    else:
        print(resp.json())

    # pull out mongo id and make it the rearrangement_id
    #newdoc = resp.json()
    #href = newdoc['_links']['rh:newdoc'][0]['href']
    #rearrangement_id = href.split('/')[-1]
    #print(rearrangement_id)
    #data = {"_id":rearrangement_id,"rearrangement_id":rearrangement_id}
    #url = 'https://' + config['api_server'] + '/meta/v3/' + config['dbname'] + '/rearrangement/' + rearrangement_id
    #print(url)
    #resp = requests.patch(url, json=data, headers=headers)
    #print(resp.status_code)
    #print(resp.text)
    #print(resp.json())
    

def getAllSubstrings(str, size=4):
    result = []
    for i in range(0, len(str)):
        for j in range(len(str), i+size-1, -1):
            result.append(str[i:j])
    return result

def parseGene(str):
    result = { "gene": None, "subgroup": None }
    aidx = str.find('*')
    if aidx < 0:
        return None
    result['gene'] = str[0:aidx]

    didx = result['gene'].find('-')
    if didx >= 0:
        result['subgroup'] = result['gene'][0:didx]
    return result

def changeGeneCall(name, name_gene, name_subgroup, obj):
    fields = obj[name].split(',')
    if len(fields) > 1:
        obj[name] = fields
        genes = []
        subgroups = []
        for i in range(0, len(fields)):
            c = fields[i]
            result = parseGene(c);
            if result == None:
                genes.append(None)
                subgroups.append(None)
            else:
                genes.append(result['gene'])
                subgroups.append(result['subgroup'])
            obj[name_gene] = genes;
            obj[name_subgroup] = subgroups;
    else:
        result = parseGene(obj[name])
        if result:
            obj[name_gene] = result['gene']
            if result['subgroup']:
                obj[name_subgroup] = result['subgroup']

# main entry
if (__name__=="__main__"):
    parser = argparse.ArgumentParser(description='Load AIRR rearrangements into VDJServer data repository.')
    parser.add_argument('load_set_start', type=int, help='Starting load set')
    parser.add_argument('repertoire_file', type=str, help='AIRR repertoire metadata file name')
    parser.add_argument('file_prefix', type=str, help='Directory prefix to find the rearrangements files')
    args = parser.parse_args()

    load_set_size = 1000

    if args:
        data = airr.load_repertoire(args.repertoire_file)
        reps = data['Repertoire']

        load_set_start = args.load_set_start
        for rep in reps:
            config = getConfig()
            token = getToken(config)
            print(token['access_token'])

            print('Loading AIRR rearrangements for repertoire: ' + rep['repertoire_id'])
            print('Starting load set: ' + str(load_set_start))
            deleteLoadSet(token, config, rep['repertoire_id'], load_set_start)
            load_set = 0

            primary_dp = None
            for dp in rep['data_processing']:
                if dp.get('primary_annotation'):
                    primary_dp = dp
            if not primary_dp:
                print('ERROR: Repertoire missing primary data processing: ' + rep['repertoire_id'])
                sys.exit(1)

            files = primary_dp['data_processing_files']
            for f in files:
                if os.path.isfile(args.file_prefix + '/' + f):
                    print('AIRR rearrangement file: ' + args.file_prefix + '/' + f)
                    reader = airr.read_rearrangement(args.file_prefix + '/' + f)
                elif os.path.isfile(args.file_prefix + '/' + primary_dp['data_processing_id'] + '/' + f):
                    print('AIRR rearrangement file: ' + args.file_prefix + '/' + primary_dp['data_processing_id'] + '/' + f)
                    reader = airr.read_rearrangement(args.file_prefix + '/' + primary_dp['data_processing_id'] + '/' + f)
                else:
                    print('ERROR: cannot find file: ' + f)
                    sys.exit(1)

                total = 0
                cnt = 0
                records = []
                for r in reader:
                    if r.get('repertoire_id') is None:
                        r['repertoire_id'] = rep['repertoire_id']
                    if len(r['repertoire_id']) == 0:
                        r['repertoire_id'] = rep['repertoire_id']
                    if r.get('data_processing_id') is None:
                        r['data_processing_id'] = primary_dp['data_processing_id']
                    if len(r['data_processing_id']) == 0:
                        r['data_processing_id'] = primary_dp['data_processing_id']
                    r['vdjserver_load_set'] = load_set
                    
                    r['receptor_id'] = r['sequence_id']
                    del r['sequence_id']

                    changeGeneCall('v_call', 'v_gene', 'v_subgroup', r)
                    changeGeneCall('d_call', 'd_gene', 'd_subgroup', r)
                    changeGeneCall('j_call', 'j_gene', 'j_subgroup', r)

                    if len(r['junction_aa']) > 3:
                        r['vdjserver_junction_substrings'] = getAllSubstrings(r['junction_aa'], 4)

                    records.append(r)
                    cnt += 1
                    total += 1
                    if cnt == load_set_size:
                        if load_set >= load_set_start:
                            print('Inserting load set: ' + str(load_set))
                            insertRearrangement(token, config, records)
                            print('Total records: ' + str(total))
                        else:
                            print('Skipping load set: ' + str(load_set))
                        cnt = 0
                        load_set += 1
                        records = []
                if cnt != 0:
                    if load_set >= load_set_start:
                        print('Inserting load set: ' + str(load_set))
                        insertRearrangement(token, config, records)
                        print('Total records: ' + str(total))
                    else:
                        print('Skipping load set: ' + str(load_set))
                    load_set += 1
                print("Total records inserted: " + str(total))
            load_set_start = 0

File: app/load/rearrangement_load.js
//
// Import data into AIRR mongo repository for a public project. This is for
// the rearrangements. This assumes you are running in the docker container.
//

'use strict';

var csv = require('csv-parser');
var fs = require('fs');
var MongoClient = require('mongodb').MongoClient;
var assert = require('assert');
var Q = require('q');
const zlib = require('zlib');

// Server environment config
var config = require('../config/config');
var mongoSettings = require('../config/mongoSettings');

// Delete all rearrangements for a repertoire_id or for
// just a given load_set.
function deleteLoadSet(repertoire_id, load_set) {
    var deferred = Q.defer();

    // get connection to database
    MongoClient.connect(mongoSettings.url, async function(err, db) {
        if (err) {
            var msg = "Could not connect to database: " + err;
            console.error(msg);
            deferred.reject(new Error(msg))
        } else {
            var v1airr = db.db(mongoSettings.dbname);
            var collection = v1airr.collection('rearrangement');

            // delete than insert repertoire
            var filter = {"repertoire_id":repertoire_id}
            if (load_set > 0)
                filter['vdjserver_load_set'] = load_set;
            console.log(filter);

            var result = await collection.deleteMany(filter);
            console.log('Deleted rearrangements: ' + result);
            db.close();
            deferred.resolve(result);
        }
    });

    return deferred.promise;
}

// Insert rearrangements for a repertoire.
// This function supports vdjserver_load_set which allows for a failed load to be
// restarted where it left off, though you need to know the last load set.
// Insertion will continue starting with the given load_set.
function insertRearrangement(repertoire_id, load_set, records) {
    var deferred = Q.defer();

    // get connection to database
    MongoClient.connect(mongoSettings.url, async function(err, db) {
        if (err) {
            var msg = "Could not connect to database: " + err;
            console.error(msg);
            deferred.reject(new Error(msg))
        } else {
            var v1airr = db.db(mongoSettings.dbname);
            var collection = v1airr.collection('rearrangement');

            // insert rearrangements
            //console.log(records[0]);
            var result = await collection.insertMany(records);
            console.log('Inserted rearrangements: ' + JSON.stringify(result['result']));
            db.close();
            deferred.resolve(result);
        }
    });

    return deferred.promise;
}

function processFile(filename, rep, load_set, load_set_start) {
    var deferred = Q.defer();

    var records = [];
    fs.createReadStream(filename)
        .pipe(zlib.createGunzip())
        .pipe(csv({separator:'\t'}))
        .on('data', (row) => {
            if (!row['repertoire_id']) row['repertoire_id'] = rep['repertoire_id'];
            if (row['repertoire_id'].length == 0) row['repertoire_id'] = rep['repertoire_id'];
            if (!row['data_processing_id'])
                row['data_processing_id'] = rep['data_processing'][0]['data_processing_id'];
            if (row['data_processing_id'].length == 0)
                row['data_processing_id'] = rep['data_processing'][0]['data_processing_id'];
            row['vdjserver_load_set'] = load_set
            //console.log(row);
            records.push(row);
            if (records.length == 10000) {
                if (load_set >= load_set_start) {
                    console.log('Inserting load set: ' + load_set);
                    //console.log('Total records: ' + total);
                    //console.log(records[0]);
                    insertRearrangement(rep['repertoire_id'], load_set, records);
                } else {
                    console.log('Skipping load set: ' + load_set);
                }
                ++load_set;
                records = [];
            }
        })
        .on('end', () => {
            if (records.length > 0) {
                if (load_set >= load_set_start) {
                    console.log('Inserting load set: ' + load_set);
                    //console.log('Total records: ' + total);
                    insertRearrangement(rep['repertoire_id'], load_set, records);
                }
            }
            console.log('AIRR TSV file successfully processed');
            deferred.resolve(load_set);
        });

    return deferred.promise;
}

console.log('Load AIRR rearrangements into VDJServer data repository.');
var myArgs = process.argv.slice(2);
if (myArgs.length != 3) {
    console.log('usage: node rearrangement_load.js load_set repertoire_file directory_prefix');
    return;
}

console.log(mongoSettings.url);

var directory_prefix = myArgs[2];
var text = fs.readFileSync(myArgs[1]);
var data = JSON.parse(text);
var reps = data['Repertoire'];

var load_set_start = parseInt(myArgs[0], 10);

(async function() {
    for (var i in reps) {
        var r = reps[i];
        if (!r['repertoire_id']) {
            console.error('Repertoire is missing repertoire_id');
            return;
        }
        if (r['repertoire_id'].length == 0) {
            console.error('Repertoire is missing repertoire_id');
            return;
        }

        var result = await deleteLoadSet(r['repertoire_id'], load_set_start);

        var primary = null;
        for (var j in r['data_processing']) {
            if (r['data_processing'][j]['primary']) {
                primary = j;
                break;
            }
        }
        if (primary == null) primary = 0;
        var files = r['data_processing'][primary]['data_processing_files'];
        console.log(files);

        var total = 0;
        var load_set = 0;
        for (var f in files) {
            console.log('AIRR rearrangement file: ' + directory_prefix + '/' + files[f]);
            load_set = await processFile(directory_prefix + '/' + files[f], r, load_set, load_set_start);
            console.log('Final load set:' + load_set);
        }

        console.log(i);
    };
})();
